### PR TITLE
[API-8] Resource Management

### PR DESCRIPTION
**Pull request by mattmess1221**
[Common](https://github.com/SpongePowered/SpongeCommon/pull/3108)

It's been nearly 3 years since I opened #1706, so here's starting fresh on a new branch with mostly new code.

The only thing I don't like is `PackDiscoverer`s. That part isn't critical to resources, so we can remove it and do it for another time if we're short on time.

I'm also not sure if we need a separate event for registering reload listeners. We may be able to get away with having them be registered during ServerAboutToStartEvent. Input is appreciated.

### CODE REVIEW

File: src/main/java/org/spongepowered/api/event/resource/PackEvent.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+         * <p>Returns a {@link MutablePackList} to expose methods for adding listeners.</p>
+         * @return The pack list
+         */
+        @Override
+        MutablePackList getPackList();
+
+        /**
+         * Register event for client resources.
+         */
+        interface Client extends RegisterPackDiscoverer {
```
Comment (Zidane) [inline src/main/java/org/spongepowered/api/event/resource/PackEvent.java]: Maybe just one event that is generic for engine?

See `StartingEngineEvent`

File: src/main/java/org/spongepowered/api/event/resource/ResourceEvent.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+         *
+         * @return The reloadable resource manager
+         */
+        @Override
+        ReloadableResourceManager getResourceManager();
+
+        /**
+         * Register event for client resources.
+         */
+        interface Client extends RegisterReloadListener {
```
Comment (Zidane) [inline src/main/java/org/spongepowered/api/event/resource/ResourceEvent.java]: Same as the above, a generic engine event.

If lifecycle, put it there along with the above.

File: src/main/java/org/spongepowered/api/event/resource/PackEvent.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+     */
+    interface RegisterPackDiscoverer extends PackEvent {
+
+        /**
+         * {@inheritDoc}
+         * <p>Returns a {@link MutablePackList} to expose methods for adding listeners.</p>
+         * @return The pack list
+         */
+        @Override
+        MutablePackList getPackList();
```
Comment (Zidane) [inline src/main/java/org/spongepowered/api/event/resource/PackEvent.java]: Do not expose the pack list, instead add a register method like the registry events.

File: src/main/java/org/spongepowered/api/event/resource/PackEvent.java [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+ * Base {@link Pack} event.
+ */
+public interface PackEvent extends Event {
+
+    /**
+     * Gets the relevant {@link PackList} for this event.
+     *
+     * @return The pack list
+     */
+    PackList getPackList();
```
Comment (Zidane) [inline src/main/java/org/spongepowered/api/event/resource/PackEvent.java]: See below.

File: src/main/java/org/spongepowered/api/event/resource/ResourceEvent.java [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+ * Base event for {@link Resource}s.
+ */
+public interface ResourceEvent extends Event {
+
+    /**
+     * Gets the relevant {@link ResourceManager} for this event.
+     *
+     * @return The resource manager
+     */
+    ResourceManager getResourceManager();
```
Comment (Zidane) [inline src/main/java/org/spongepowered/api/event/resource/ResourceEvent.java]: Do not expose the manager, expose register methods.

File: src/main/java/org/spongepowered/api/event/resource/ResourceEvent.java (hunk 2 above)
Comment (mattmess1221) [reply src/main/java/org/spongepowered/api/event/resource/ResourceEvent.java]: I wouldn't consider it a lifecycle event. It's just a specific event that does a specific task once at a specific time during init.

**Comment by Zidane** [comment]
Closing in favor of https://github.com/SpongePowered/SpongeAPI/tree/feature/resources

Thank you for all your work on this Killjoy.

**Pull request closed without merging**